mysys_ssl/my_crypt.cc

```diff
@@ -0,0 +1,378 @@
+
+#include <my_crypt.h>
+
+#include "boost/move/unique_ptr.hpp"
+
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+#define ERR_remove_state(X) ERR_clear_error()
+#else
+#define EVP_CIPHER_CTX_buf_noconst(ctx) ((ctx)->buf)
+#define RAND_OpenSSL() RAND_SSLeay()
+#endif
+
+#define make_aes_dispatcher(mode)                               \
+  static inline const EVP_CIPHER *aes_ ## mode(uint klen)       \
+  {                                                             \
+    switch (klen) {                                             \
+    case 16: return EVP_aes_128_ ## mode();                     \
+    case 24: return EVP_aes_192_ ## mode();                     \
+    case 32: return EVP_aes_256_ ## mode();                     \
+    default: return 0;                                          \
+    }                                                           \
+  }
+
+make_aes_dispatcher(ecb)
+make_aes_dispatcher(cbc)
+#ifdef HAVE_EncryptAes128Ctr
+make_aes_dispatcher(ctr)
+#endif /* HAVE_EncryptAes128Ctr */
+#ifdef HAVE_EncryptAes128Gcm
+make_aes_dispatcher(gcm)
+#endif
+
+const EVP_CIPHER *(*ciphers[])(uint)= {
+    aes_ecb, aes_cbc
+#ifdef HAVE_EncryptAes128Ctr
+  , aes_ctr
+#ifdef HAVE_EncryptAes128Gcm
+  , aes_gcm
+#endif
+#endif
+};
+
+struct MyEncryptionCTX::Impl
+{
+  Impl() 
+  : ctx(NULL)
+  {}
+
+  EVP_CIPHER_CTX *ctx;
+};
+
+MyEncryptionCTX::MyEncryptionCTX()
+{
+  pimpl = new Impl();
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+  pimpl->ctx= new EVP_CIPHER_CTX();
+  EVP_CIPHER_CTX_init(pimpl->ctx);
+#else
+  pimpl->ctx= EVP_CIPHER_CTX_new();
+#endif
+}
+
+MyEncryptionCTX::~MyEncryptionCTX()
+{
+  
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+  EVP_CIPHER_CTX_cleanup(pimpl->ctx);
+  delete pimpl->ctx;
+#else
+  EVP_CIPHER_CTX_reset(pimpl->ctx);
+  EVP_CIPHER_CTX_free(pimpl->ctx);
+#endif
+  ERR_remove_state(0);
+  delete pimpl;
+}
+
+int MyEncryptionCTX::init(const my_aes_mode mode, int encrypt, const uchar *key,
+                size_t klen, const uchar *iv, size_t ivlen)
+{
+  compile_time_assert(MY_AES_CTX_SIZE >= sizeof(MyEncryptionCTX));
+  if (unlikely(!ciphers[mode](klen)))
+    return MY_AES_BAD_KEYSIZE;
+
+  if (!EVP_CipherInit_ex(pimpl->ctx, ciphers[mode](klen), NULL,
+                         key, iv, encrypt))
+    return MY_AES_OPENSSL_ERROR;
+
+  DBUG_ASSERT(EVP_CIPHER_CTX_key_length(pimpl->ctx) == (int)klen);
+  DBUG_ASSERT(EVP_CIPHER_CTX_iv_length(pimpl->ctx) <= (int)ivlen);
+
+  return MY_AES_OK;
+}
+
+int MyEncryptionCTX::update(const uchar *src, size_t slen, uchar *dst, size_t *dlen)
+{
+  if (!EVP_CipherUpdate(pimpl->ctx, dst, (int*)dlen, src, slen))
+    return MY_AES_OPENSSL_ERROR;
+  return MY_AES_OK;
+}
+
+int MyEncryptionCTX::finish(uchar *dst, size_t *dlen)
+{
+  if (!EVP_CipherFinal_ex(pimpl->ctx, dst, (int*)dlen))
+    return MY_AES_BAD_DATA;
+  return MY_AES_OK;
+}
+
+class MyEncryptionCTX_nopad : public MyEncryptionCTX
+{
+public:
+  const uchar *key;
+  uint klen, buf_len;
+  uchar oiv[MY_AES_BLOCK_SIZE];
```

Comment:
Why not to store `oiv` here as just a pointer `const uchar* oiv`?
2 places where `MyEncryptionCTX::init()` is called with `ENCRYPTION_FLAG_NOPAD`:
1. in `encrypt_event()` it outlives the call.
2. in `Event_encrypter::init()` it can be bound to `crypto->iv`.

Reply:
It is much safer this way, you do not have keep this memory allocated and it would be weird to use such function. Besides, iv in Event_encrypter::init() is a local variable that would have to be allocated on heap. It is small array and imho if possible it should be allocated on stack.